content/cpi-api-v2.md

```diff
@@ -0,0 +1,115 @@
+
+```yaml
+resource_pools:
+- name: large_machines
+  cloud_properties: {instance_type: r3.8xlarge}
+```
+
+## Methods
+
+- All V1 contracts must still be supported. See [CPI API V1](cpi-api-v1.md).
+- To differentiate V2 calls the caller needs to pass in `"api_version": 2` in the context
+
+### Reference Table (Based on each component version)
+* Reg./reg. : Registry
+
+| Director | CPI | Stemcell  | Should update Reg.   | Add agent setting to Iaas `user-metadata`   |
+|----------|-----|-----------|----------------------|---|
+| 1  | 1  | 1  | Update Reg.  | Dont add agent setting to iaas  |
+| 1  | 1  | 2  | Update Reg.  | Dont add agent setting to iaas  |
+| 1  | 2  | 2  | Update Reg.  | Dont add agent setting to iaas  |
+| 2  | 2  | 2  | DON'T add anything to reg.   | Yes, agent will read `user-metadata` and not call reg.  |
+| 1  | 2  | 1  | Update Reg.  | Dont add agent setting to iaas  |
+| 2  | 2  | 1  | Update Reg.  | Dont add agent setting to iaas (no agent support)  |
+| 2  | 1  | 1  | Update Reg. (cpi will by default update reg.)  | Dont add agent setting to iaas |
+| 2  | 1  | 2  | Update Reg. (cpi will by default update reg.)  | Dont add agent setting to iaas |
+
+### Implementation
+
+API version 2 of CPIs will differ from version 1 by the following:
+
+ - Director will send cpi api_version based on info response from for all CPI calls
+ - Director will also send stemcell `api_version` for all CPI calls
+ - Director will be expecting V2 responses IF
+   - CPI info call provide max supported version as >=2 AND
+   - Stemell api_version is >=2 AND
+   - `api_version` 2 is requested in the call
+
+
+  ```json
+   "context": {
+    "director_uuid": "<director-uuid>",
+    "request_id": "<cpi-request-id>",
+    "vm": {
+      "stemcell": {
+        "api_version": 2
+      }
+    }
+  },
+  "api_version": 2
+  ```
+
+#### Stemcell changes with V2 contracts:
+
+- `stemcell.MF` should contain `api_version` if it contains a v2 supported agent in it, so that director, cpi and cli can support not usage of registry.
+	-  if api_version is not provided in the `stemcell.MF` it treats it as api_version 1
+	-  in order to get new behaviour `api_version > 1`
+
+```yaml
+---
+#### stemcell api_version
+api_version: 2
+name: bosh-aws-xen-hvm-ubuntu-trusty-go_agent
+version: '3546.14'
+bosh_protocol: '1'
+sha1: c186de6ef6e034bc93513440b9071b5f4696fa32
+operating_system: ubuntu-trusty
+stemcell_formats:
+- aws-light
+cloud_properties:
+  ami:
+    us-east-1: ami-xxxxxx
+    us-west-1: ami-xxxxxx
+```
+
+### Changes in V2 contracts
+
+ * [info](cpi-api-v2-method/info.md)
+ * VM Management
+    * [create_vm](cpi-api-v2-method/create-vm.md)
+    * [delete_vm](cpi-api-v2-method/delete-vm.md)
+ * Disk Management
+    * [attach_disk](cpi-api-v2-method/attach-disk.md)
+    * [detach_disk](cpi-api-v2-method/detach-disk.md)
+
+
+#### Additional methods in V1:
```

Comment:
I'm not sure what the semantics of 'additional methods' in an API version are. Are you saying those should are optional? I think API methods either are part of an API version or not.

Reply:
We are working on this PR right now in order to improve the content and structure.

Reply:
We have updated the title for that section.